filebeat/input/filestream/internal/input-logfile/update_writer.go

```diff
@@ -0,0 +1,125 @@
+package input_logfile
+
+import (
+	"context"
+	"sync"
+
+	"github.com/elastic/go-concert/ctxtool"
+	"github.com/elastic/go-concert/unison"
+)
+
+// updateWriter asynchronously writes all updates to the persistent store.
+// All updates are tracked as key value pairs. In case of back-pressure due to the local disk
+// we overwrite the pending states that have not been written in memory,
+// until the disk is ready for more updates.
+type updateWriter struct {
+	store *store
+	tg    unison.TaskGroup
+
+	// we use a chan as conditional, so we can break on context cancellation.
+	// `waiter` is set if the writer is waiting for new entries to be reported to the store.
+	mutex  sync.Mutex
+	waiter chan struct{}
+
+	// pending update operations for key value pairs. The new state always
+	// overwrites the old state.
+	pending map[string]int
+	updates []scheduledUpdate
+}
+
+type scheduledUpdate struct {
+	op *updateOp
+	n  uint
+}
+
+func newUpdateWriter(store *store) *updateWriter {
+	w := &updateWriter{
+		store:   store,
+		pending: map[string]int{},
+	}
+	w.tg.Go(func(ctx unison.Canceler) error {
+		w.run(ctxtool.FromCanceller(ctx))
+		return nil
+	})
+
+	return w
+}
+
+// Close stops the background writing provess and attempts to serialize
+// all pending operations.
+func (w *updateWriter) Close() {
+	w.tg.Stop()
+	w.syncStates(w.updates)
+}
+
+// Set overwrites key value pair in the pending update operations.
+func (w *updateWriter) Schedule(op *updateOp, n uint) {
+	w.mutex.Lock()
+	defer w.mutex.Unlock()
```

Comment:
`Schedule` is a point of contention here. But it is still better then blocking and waiting for the serialization and IO to be finished.